google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -319,52 +373,359 @@ private boolean isAlive() {
+          subscriberStub
+              .modifyAckDeadlineCallable()
+              .futureCall(
+                  ModifyAckDeadlineRequest.newBuilder()
+                      .setSubscription(subscription)
+                      .addAllAckIds(ackIdsInRequest)
+                      .setAckDeadlineSeconds(modackRequestData.getDeadlineExtensionSeconds())
+                      .build());
+      ApiFutures.addCallback(modackFuture, callback, directExecutor());
+      pendingOperations++;
+    }
     ackOperationsWaiter.incrementPendingCount(pendingOperations);
   }
+
+  private Map<String, String> getMetadataMapFromThrowable(Throwable t) {
+    // This converts a Throwable (from a "OK" grpc response) to a map of metadata
+    // will be of the format:
+    // {
+    //    "ACK-ID-1": "PERMANENT_*",
+    //    "ACK-ID-2": "TRANSIENT_*"
+    // }
+    com.google.rpc.Status status = StatusProto.fromThrowable(t);
+    Map<String, String> metadataMap = new HashMap<>();
+    if (status != null) {
+      for (Any any : status.getDetailsList()) {
+        if (any.is(ErrorInfo.class)) {
+          try {
+            ErrorInfo errorInfo = any.unpack(ErrorInfo.class);
+            metadataMap = errorInfo.getMetadataMap();
+          } catch (Throwable throwable) {
+          }
+        }
+      }
+    }
+    return metadataMap;
+  }
+
+  private ApiFutureCallback<Empty> getCallback(
+      List<AckRequestData> ackRequestDataList,
+      int deadlineExtensionSeconds,
+      boolean isAckRequest,
+      long currentBackoffMillis) {
+    // This callback handles retries, and sets message futures
+    return new ApiFutureCallback<Empty>() {
+      @Override
+      public void onSuccess(Empty empty) {
+        ackOperationsWaiter.incrementPendingCount(-1);
+        for (AckRequestData ackRequestData : ackRequestDataList) {
+          // This will check if a response is needed, and if it has already been set
+          ackRequestData.setAckResponse(AckResponse.SUCCESSFUL);
+          // Remove from our pending operations
+          pendingRequests.remove(ackRequestData);
+        }
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        // Remove from our pending operations
+        ackOperationsWaiter.incrementPendingCount(-1);
+        Map<String, String> metadataMap = getMetadataMapFromThrowable(t);
+        List<AckRequestData> ackRequestDataArrayRetryList = new ArrayList<AckRequestData>();
+        ackRequestDataList.forEach(
+            ackRequestData -> {
+              String ackId = ackRequestData.getAckId();
+              if (metadataMap.containsKey(ackId)) {
+                // An error occured
+                String errorMessage = metadataMap.get(ackId);
+                // Make sure the message has not already been handled
+                if ((ackRequestData.getMessageFuture() != null)
+                    && ackRequestData.getMessageFuture().isDone()) {
+                  logger.log(
+                      Level.WARNING,
```

Comment:
Does this need to be a warning?

Reply:
This was changed, not sure what the context was. Let me know if this is still an issue.